external-crates/move/crates/move-analyzer/src/symbols.rs

```diff
@@ -2193,39 +2539,219 @@ impl Symbolicator {
+            &def_fhash,
+        );
+        let ud = UseDef::new(
+            references,
+            use_pos.file_hash(),
+            name_start,
+            def_fhash,
+            def.name_start,
+            use_name,
+            on_hover,
+            ident_type_def_loc,
+            doc_string,
+        );
+        use_defs.insert(name_start.line, ud.clone());
+        return Some(ud);
+    }
+    None
+}
 
-    fn on_hover_to_type_def_loc(&self, on_hover: &IdentOnHover) -> Option<DefLoc> {
-        match on_hover {
-            IdentOnHover::Type(t) => self.type_def_loc(t),
-            IdentOnHover::Function(_, _, _, _, _, _, ret) => self.type_def_loc(ret),
-            IdentOnHover::Struct(mod_ident, name, _, _, _) => self.find_struct(mod_ident, name),
-            IdentOnHover::Field(_, _, _, t) => self.type_def_loc(t),
-            IdentOnHover::Local(_, t, _) => self.type_def_loc(t),
-            IdentOnHover::Const(_, t, _) => self.type_def_loc(t),
+fn on_hover_to_type_def_loc(
+    mod_outer_defs: &BTreeMap<String, ModuleDefs>,
+    on_hover: &IdentOnHover,
+) -> Option<DefLoc> {
+    match on_hover {
+        IdentOnHover::Type(t) => type_def_loc(mod_outer_defs, t),
+        IdentOnHover::Function(_, _, _, _, _, _, ret) => type_def_loc(mod_outer_defs, ret),
+        IdentOnHover::Struct(mod_ident, name, _, _, _) => {
+            find_struct(mod_outer_defs, mod_ident, name)
         }
+        IdentOnHover::Field(_, _, _, t) => type_def_loc(mod_outer_defs, t),
+        IdentOnHover::Local(_, t, _) => type_def_loc(mod_outer_defs, t),
+        IdentOnHover::Const(_, t, _) => type_def_loc(mod_outer_defs, t),
+        IdentOnHover::Module(_) => None,
     }
+}
 
-    fn type_def_loc(&self, sp!(_, t): &Type) -> Option<DefLoc> {
-        match t {
-            Type_::Ref(_, r) => self.type_def_loc(r),
-            Type_::Apply(_, sp!(_, TypeName_::ModuleType(sp!(_, mod_ident), struct_name)), _) => {
-                self.find_struct(mod_ident, &struct_name.value())
+fn type_def_loc(mod_outer_defs: &BTreeMap<String, ModuleDefs>, sp!(_, t): &Type) -> Option<DefLoc> {
+    match t {
+        Type_::Ref(_, r) => type_def_loc(mod_outer_defs, r),
+        Type_::Apply(_, sp!(_, TypeName_::ModuleType(sp!(_, mod_ident), struct_name)), _) => {
+            find_struct(mod_outer_defs, mod_ident, &struct_name.value())
+        }
+        _ => None,
+    }
+}
+
+fn find_struct(
+    mod_outer_defs: &BTreeMap<String, ModuleDefs>,
+    mod_ident: &ModuleIdent_,
+    struct_name: &Symbol,
+) -> Option<DefLoc> {
+    let mod_defs = match mod_outer_defs.get(&format!("{}", mod_ident)) {
+        Some(v) => v,
+        None => return None,
+    };
+    mod_defs.structs.get(struct_name).map(|struct_def| {
+        let fhash = mod_defs.fhash;
+        let start = struct_def.name_start;
+        DefLoc { fhash, start }
+    })
+}
+
+/// Extracts the docstring (/// or /** ... */) for a given definition by traversing up from the line definition
```

Comment:
Worth adding a test for the `/** ... */` syntax as well?

Reply:
Present and accounted for! (unless I am misunderstanding the question). In source files they start here: https://github.com/MystenLabs/sui/blob/a0ee03e9af815ebedf653e2c2492edc358881eb6/external-crates/move/crates/move-analyzer/tests/symbols/sources/M6.move#L20 and in the test itself here: https://github.com/MystenLabs/sui/blob/a0ee03e9af815ebedf653e2c2492edc358881eb6/external-crates/move/crates/move-analyzer/src/symbols.rs#L2835